File: testing/compose/js/chatbot/src/tools/catalogue.ts
export type ToolName =
  | "web_search"
  | "code_executor"
  | "document_reader"
  | "data_analyst"
  | "plan_reviewer"
  | "risk_checker"
  | "summarizer"
  | "diff_generator";

export interface ToolContribution {
  tool_call_count: number;
  workflow_steps: number;
  tokens_prompt: number;
  tokens_completion: number;
  artifact_count: number;
  output_size_proxy: number;
}

export interface ToolTraceEntry {
  tool_name: ToolName;
  invoked_at_ms: number;
  output_size: number;
}

export const TOOL_CONTRIBUTIONS: Record<ToolName, ToolContribution> = {
  web_search: {
    tool_call_count: 1,
    workflow_steps: 1,
    tokens_prompt: 500,
    tokens_completion: 0,
    artifact_count: 0,
    output_size_proxy: 0,
  },
  code_executor: {
    tool_call_count: 1,
    workflow_steps: 1,
    tokens_prompt: 0,
    tokens_completion: 0,
    artifact_count: 1,
    output_size_proxy: 0.5,
  },
  document_reader: {
    tool_call_count: 1,
    workflow_steps: 1,
    tokens_prompt: 2000,
    tokens_completion: 0,
    artifact_count: 0,
    output_size_proxy: 0,
  },
  data_analyst: {
    tool_call_count: 1,
    workflow_steps: 0,
    tokens_prompt: 0,
    tokens_completion: 800,
    artifact_count: 1,
    output_size_proxy: 1,
  },
  plan_reviewer: {
    tool_call_count: 1,
    workflow_steps: 1,
    tokens_prompt: 0,
    tokens_completion: 0,
    artifact_count: 0,
    output_size_proxy: 0,
  },
  risk_checker: {
    tool_call_count: 1,
    workflow_steps: 0,
    tokens_prompt: 0,
    tokens_completion: 0,
    artifact_count: 0,
    output_size_proxy: 0,
  },
  summarizer: {
    tool_call_count: 1,
    workflow_steps: 0,
    tokens_prompt: 0,
    tokens_completion: 0,
    artifact_count: 1,
    output_size_proxy: 0.8,
  },
  diff_generator: {
    tool_call_count: 1,
    workflow_steps: 0,
    tokens_prompt: 0,
    tokens_completion: 0,
    artifact_count: 1,
    output_size_proxy: 0.3,
  },
};

export function functionSchemas(activeTools: ToolName[]): object[] {
  return activeTools.map((name) => ({
    type: "function",
    function: {
      name,
      description: `Synthetic ${name} tool used by the ai-heeczer local stack.`,
      parameters: {
        type: "object",
        properties: {
          input: { type: "string" },
        },
        required: ["input"],
      },
    },
  }));
}

export function traceForTools(tools: ToolName[]): ToolTraceEntry[] {
  return tools.map((toolName, index) => ({
    tool_name: toolName,
    invoked_at_ms: index * 25,
    output_size: TOOL_CONTRIBUTIONS[toolName].output_size_proxy,
  }));
}

File: testing/compose/go/chatbot/tools/catalogue.go
package tools

// ToolTraceEntry is displayed by the local chatbot UI and mirrored into event metadata.
type ToolTraceEntry struct {
	ToolName    string  `json:"tool_name"`
	InvokedAtMs int     `json:"invoked_at_ms"`
	OutputSize  float64 `json:"output_size"`
}

var contributions = map[string]float64{
	"web_search":      0.0,
	"code_executor":   0.5,
	"document_reader": 0.0,
	"data_analyst":    1.0,
	"plan_reviewer":   0.0,
	"risk_checker":    0.0,
	"summarizer":      0.8,
	"diff_generator":  0.3,
}

// TraceForTools turns a mock script's tool list into a stable trace.
func TraceForTools(activeTools []string) []ToolTraceEntry {
	trace := make([]ToolTraceEntry, 0, len(activeTools))
	for index, toolName := range activeTools {
		trace = append(trace, ToolTraceEntry{
			ToolName:    toolName,
			InvokedAtMs: index * 25,
			OutputSize:  contributions[toolName],
		})
	}
	return trace
}

// FunctionSchemas returns OpenAI-compatible function declarations.
func FunctionSchemas(activeTools []string) []map[string]any {
	schemas := make([]map[string]any, 0, len(activeTools))
	for _, toolName := range activeTools {
		schemas = append(schemas, map[string]any{
			"type": "function",
			"function": map[string]any{
				"name":        toolName,
				"description": "Synthetic " + toolName + " tool used by the ai-heeczer local stack.",
				"parameters": map[string]any{
					"type":       "object",
					"properties": map[string]any{"input": map[string]any{"type": "string"}},
					"required":   []string{"input"},
				},
			},
		})
	}
	return schemas
}
